[16.0][MIG] l10n_be_iso20022_pain

sbejaoui opened the pull request:
needs: https://github.com/OCA/bank-payment/pull/1314

sbidoul committed:
[ADD] l10n_be_iso20022_pain

yvaucher committed:
Add bug tracker link on README.rst

lmignon committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

ThomasBinsfeld committed:
[MIG] 10n_be_iso20022_pain to 9.0

pedrobaeza committed:
[MIG] Rename manifest files

ThomasBinsfeld committed:
[FIX] l10n_be_iso20022: translation of selection field

ThomasBinsfeld committed:
[ADD] l10_be_iso20022_pain: french translation

ThomasBinsfeld committed:
[MIG] l10n_be_iso20022_pain to 10.0

sbidoul committed:
[FIX] header underline in readme

sbidoul committed:
[FIX] l10n_be_iso20022_pain: test was really broken, fixed

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
remove obsolete .pot files [ci skip]

oca-travis committed:
[UPD] Update l10n_be_iso20022_pain.pot

OCA-git-bot committed:
[ADD] icon.png

ThomasBinsfeld committed:
[MIG] l10n_be_iso20022_pain: run pre-commit

ThomasBinsfeld committed:
[MIG] l10n_be_iso20022_pain to 13.0

Commit:
[UPD] Update l10n_be_iso20022_pain.pot

OCA-git-bot committed:
l10n_be_iso20022_pain 13.0.1.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-belgium-13.0/l10n-belgium-13.0-l10n_be_iso20022_pain
Translate-URL: https://translation.odoo-community.org/projects/l10n-belgium-13-0/l10n-belgium-13-0-l10n_be_iso20022_pain/

sbejaoui committed:
[16.0][MIG] l10n_be_iso20022_pain

rousseldenis commented on l10n_be_iso20022_pain/models/account_move.py:
```diff
@@ -25,7 +25,8 @@ class AccountMove(models.Model):
 
     reference_type = fields.Selection(
-        selection_add=[("bba", "BBA Structured Communication")]
+        selection_add=[("bba", "BBA Structured Communication")],
```
@sbejaoui Is this still needed following this:

https://github.com/odoo/odoo/blob/ec5f6003f8f6f8de16c6c8b2b023a0511a55d498/addons/account/models/account_move.py#L2605

And this:

https://github.com/odoo/odoo/blob/ec5f6003f8f6f8de16c6c8b2b023a0511a55d498/addons/l10n_be/models/account_move.py#L20C9-L20C42

sbejaoui replied:
Yes, in the previous versions, `_get_invoice_reference_be_partner` was located in `l10n_be_invoice_bba`, but now it has been merged into `l10n_be`. This method provides a default BBA-formatted reference.

`l10n_be_iso20022_pain` add a new reference type `bba` to ensure that the reference (whether generated or edited by the user) complies with the valid BBA format.

sbidoul replied:
I don't think we should add a reference_type. If `reference_type` is `structured` and `journal_id.invoice_reference_model` is `be` it means we have a bba structured communication.

sbejaoui replied:
Correct, 

I believe we should keep only the check for the ref in this module.

cc @ThomasBinsfeld

luc-demeyer commented on l10n_be_iso20022_pain/models/account_move.py:
```diff
@@ -0,0 +1,70 @@
+    _inherit = "account.move"
+
+    reference_type = fields.Selection(
+        selection_add=[("bba", "BBA Structured Communication")],
```
"bba" in payment order should be "BBA" (uppercase)

sbidoul replied:
@ThomasBinsfeld can you elaborate why you downvoted the comment above?
@luc-demeyer if you have a link to the spec of what must go in a SEPA Credit Transfer file for BBA structured communications, that would be helpful?

ThomasBinsfeld replied:
I didn't see the need to change that and imply some refactoring. But this part has been removed since.

ThomasBinsfeld commented:
LGTM

sbejaoui committed:
[IMP] remove bba option from communication_type on payment line

sbidoul commented:
@sbejaoui @ThomasBinsfeld Is it not necessary to check and/or enforce `invoice_reference_model = 'be'`, for instance in the migration script or in `_check_communication` ?

ThomasBinsfeld commented on l10n_be_iso20022_pain/models/account_payment_line.py:
```diff
@@ -0,0 +1,19 @@
+class AccountPaymentLine(models.Model):
+    _inherit = "account.payment.line"
+
+    @api.constrains("communication", "communication_type")
+    def _check_communication(self):
+        for rec in self:
+            if rec.communication_type == "structured" and not check_bbacomm(
+                rec.communication
+            ):
+                raise ValidationError(_("Invalid BBA Structured Communication !"))
```
```suggestion
    @api.constrains("order_id", "communication", "communication_type")
    def _check_communication(self):
        for rec in self:
            if (
                rec.order_id.journal_id.invoice_reference_model == "be"
                and rec.communication_type == "structured"
                and not check_bbacomm(rec.communication
            ):
                raise ValidationError(_("Invalid BBA Structured Communication !"))
```

sbejaoui replied:
@ThomasBinsfeld , I made the field invoice_reference_model visible for the bank journal so we can perform this check. See the last commit.

ThomasBinsfeld commented:
> @sbejaoui @ThomasBinsfeld Is it not necessary to check and/or enforce `invoice_reference_model = 'be'`, for instance in the migration script or in `_check_communication` ?

The migration script looks to me. But the constraint on `account.payment.line` should check the reference model of the journal indeed.

sbejaoui committed:
[IMP] l10n_be_iso20022_pain: check BBA structured communication on payment line

check BBA structured communication on payment line only if  the journal invoice reference model is "be"

Co-authored-by: Thomas Binsfeld (ACSONE) <thomas.binsfeld@acsone.eu>

sbidoul commented on l10n_be_iso20022_pain/views/account_journal.xml:
```diff
@@ -0,0 +1,23 @@
+            >
+                <attribute
+                    name="attrs"
+                >{'invisible': [('type', 'not in', ('sale', 'bank'))]}</attribute>
```
I'm not sure it makes sense to make invoice_reference_model and invoice_reference_type visible for other journals than customer invoice journals. Because when we pay invoice bills with a journal, we don't control the type of reference we need to use for the suppliers. So I think invoice_reference_model and invoice_reference_type should only be used when generating payment reference on documents that we generate, not documents that we receive from suppliers.

sbejaoui replied:
In this case, we remove the check on the payment reference in payment orders. We assume that the reference was checked for our invoices, and we don't concern ourselves with whether it's correct for vendor bills.

sbidoul commented on l10n_be_iso20022_pain/models/account_move.py:
```diff
@@ -0,0 +1,40 @@
+                and rec.journal_id.invoice_reference_type == "invoice"
+                and rec.reference_type == "structured"
+            ):
+                return check_bbacomm(rec.ref)
```
Should we not use `rec.payment_reference` here?

sbidoul replied:
At least for vendor bills, where `ref` is the supplier invoice number?

sbejaoui replied:
> Should we not use `rec.payment_reference` here?

yes

sbejaoui replied:
> At least for vendor bills, where `ref` is the supplier invoice number?

only for out_invoice, and don't check for VB

remytms replied:
@sbejaoui If you agree with @sbidoul on the `rec.payment_reference` can you make the changes ? :)

remytms commented:
/ocabot migration l10n_be_iso20022_pain

remytms commented:
If no reaction, the PR will be marked as stale.
